### PR TITLE
Correctly parse avoid_polygons

### LINKED ISSUE #118: ORS Tools avoid polygon
Hi,
I am trying to use ORS Tools to avoid a polygon representing an area I want to route around.

I keep getting the following error message:

AttributeError: 'QgsFeature' object has no attribute 'transform' 
Traceback (most recent call last):
  File "C:/Users/User/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\ORStools\gui\ORStoolsDialog.py", line 273, in run_gui_control
    params = directions.get_parameters()
  File "C:/Users/User/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\ORStools\gui\directions_gui.py", line 107, in get_parameters
    geom.transform(transformer)
AttributeError: 'QgsFeature' object has no attribute 'transform'

I am new to QGIS and might be making a rookie mistake. (Am I creating the polygon layer correctly?)
Any help/advise would be highly appreciated.

Regards,
Jocha.

### PR DESCRIPTION
**Pull request by koebi**
Fixes #118

### CODE REVIEW

File: ORStools/gui/directions_gui.py [hunk 1]
```diff
@@ -31,11 +31,44 @@
 
 from PyQt5.QtWidgets import QCheckBox
 
-from ORStools.utils import transform
+from ORStools.utils import transform, exceptions
+
+
+def _get_avoid_polygons(layer):
```
Comment (koebi) [inline ORStools/gui/directions_gui.py:37]: why is this an internal function and not a method of `Directions` as before?

File: ORStools/gui/directions_gui.py [hunk 2]
```diff
@@ -31,11 +31,44 @@
...
+        if feature.hasGeometry():
+            geom = feature.geometry()
+            geom.transform(transformer)
+            if polygons is None:
+                polygons = geom
+            else:
+                polygons = polygons.combine(geom)
+
+    if not polygons:
+        raise exceptions.GenericServerError(
```
Comment (koebi) [inline ORStools/gui/directions_gui.py:61]: I don't think the code should raise an exception here.
Either we want this to fail w/ an API Error, or we want to warn our users in the same way that we do when points or the api-key is missing.

**Review by koebi — COMMENTED** [review]
Two questions, see line comments :)

File: ORStools/gui/directions_gui.py (hunk 2 above)
Comment (TheGreatRefrigerator) [reply ORStools/gui/directions_gui.py]: Changed to info popup and request abortion strategy

File: ORStools/gui/directions_gui.py (hunk 1 above)
Comment (TheGreatRefrigerator) [reply ORStools/gui/directions_gui.py]: This was moved to class method of `Directions` again as it needed the parent `self.dlg`.

File: ORStools/gui/directions_gui.py [hunk 3]
```diff
@@ -135,14 +103,50 @@ def get_parameters(self):
...
+            if feature.hasGeometry():
+                geom = feature.geometry()
+                geom.transform(transformer)
+                if polygons is None:
+                    polygons = geom
+                else:
+                    polygons = polygons.combine(geom)
+
+        if not polygons:
+            QMessageBox.warning(
```
Comment (koebi) [inline ORStools/gui/directions_gui.py:138]: I think we should issue this warning box in `ORStoolsDialog` as QMessageBox is already imported there.
We can then just return polygons here.

File: ORStools/gui/directions_gui.py [hunk 4]
```diff
@@ -135,14 +103,50 @@ def get_parameters(self):
         if self.dlg.avoidpolygon_group.isChecked():
             layer = self.dlg.avoidpolygon_dropdown.currentLayer()
             if layer:
-                polygons = _get_avoid_polygons(layer)
+                polygons = self._get_avoid_polygons(layer)
                 self.options['avoid_polygons'] = polygons
```
Comment (koebi) [inline ORStools/gui/directions_gui.py:107]: Maybe we should load via `json.loads(polygons.asJson())` here? I'm unsure whether this behaves correctly if `polygons` is `None` though…

**Pull request merged**